### DISCUSSION TITLE
Find the status of a button or if it's changed?

### DISCUSSION THREAD

**Discussion by jpmorrison** [mobile-dev-inc/Maestro#1620] [Q&A]
I can tapOn a button and I think I can assertVisible on the button text (Pool) but I want to find status of the button.
Is there any way to do this without access to the app source code?

The button is movable and it might be missing or there could be other buttons representing other devices. 
I tried tapOn "Pool" but it doesn't work. It would be really nice to lookup the text to find the point: 13%,42% 

![image](https://github.com/mobile-dev-inc/maestro/assets/1448696/6676c5d5-98d5-42db-a40e-95ae62c7033b)
![image](https://github.com/mobile-dev-inc/maestro/assets/1448696/ddec678f-7f49-4ef5-a6fb-95338fc6c799)

**Comment by portyu9** [comment]
Hi @jpmorrison 

Current Maestro can check several functional states directly with **state selectors**:

```yaml
- assertVisible:
    text: "Pool"
    enabled: true
```

Available state selectors include:

- `enabled`
- `checked`
- `focused`
- `selected`

You also should not normally need to convert `"Pool"` into an absolute coordinate. `tapOn` resolves the element's current position each time:

```yaml
- tapOn:
    text: "Pool"
```

so moving the button around between devices is fine as long as it remains exposed through the accessibility hierarchy.

If by "status" you mean something purely visual — for example the button changes color but exposes no corresponding accessibility state/text — Maestro cannot infer an arbitrary semantic status from the color. In that case, ideally expose that state through accessibility semantics/text, or use screenshot assertions if visual appearance itself is what you need to test.

If `tapOn: "Pool"` cannot find the button, inspect the hierarchy in Maestro Studio first; that usually means `"Pool"` is not actually exposed as a separate accessible element.

Please mark this answer as accepted if it helped, thank you.
